### ISSUE TITLE
Interrupt function for matrix not working

### ISSUE THREAD

**Issue by mkandulavm** [valhalla/valhalla#4583]
Hi

I have tested the interrupt function for actor->route and it functions as expected.

However, for the request made to actor->matrix, the interrupt is not getting queried/called and thus does not exit for the condition being enabled.

May I know if this is a known issue ? Kindly let me know if you need more details.

**Comment by nilsnolde** [comment]
Are you on master?

**Comment by nilsnolde** [comment]
Ah well maybe we forgot to add it to the actor

**Comment by mkandulavm** [comment]
> Are you on master?
yes.

**Comment by mkandulavm** [comment]
can I know where the call happens for checking with the interrupt function ? I can debug and provide feedback if this is actually a bug or something I missed. Thank you.

**Comment by nilsnolde** [comment]
I checked and it's on the actor interface.. however, the whole `interrupt` thing is _very_ all over the place, it's being set on pretty much every worker, algorithm etc. the actor interface has a spot for you to provide an interrupt function, but without looking into it more, no idea how to leverage that..

fwiw, it's (trying to be) set on the matrix algos:
https://github.com/valhalla/valhalla/blob/efff65d35c68c9198d504e07087d3ae2b71e5e1f/src/thor/matrix_action.cc#L110

**Comment by kevinkreiser** [comment]
i wrote it so i'll explain.

the interrupt is a mechanism to allow for aborting work that is in progress. when you are running a service that serves tons and tons of requests of different types, its advantageous to be able to abort work on a request if the client who requested it no longer wants it.

this works through the prime_server api but it could work for other mechanisms of external control over the running of valhalla. anyway, when the server hears a client no longer cares about a request (it gets a disconnect), it sends a message over an interrupt socket which all workers subscribe to and receive. when a worker does work it calls into the application code that is registered with the worker itself. when it does this it also passes a callback function, this is the interrupt function.

the application is meant to call this function periodically while it is doing its work just to check if it should abort the work. as long as all the tight loops in your application make sure to check this call back the iterrupt mechanism should work.

basically the callback checks this interrupt socket and if they are currently working on the request that was cancelled the interrupt callback will throw an interrupt_t exception.

if you are finding that the interrupt isnt working properly there are two things to check:

1. are all the tight loops in valhalla actually calling the interrupt callback periodically. i know that when i initially implemented it i didnt put it in every single place and when we add new places its easy to forget to do it.
2. is your service configured properly. check that the workers and server both have the same interrupt socket configured

**Comment by nilsnolde** [comment]
> are all the tight loops in valhalla actually calling the interrupt callback periodically. i know that when i initially implemented it i didnt put it in every single place and when we add new places its easy to forget to do it.

for matrix I can vouch this is now true (but true, it wasn't until recently, that's why I asked if master was used).

I'll have to look at it again, but I previously tried for a bit to follow the interrupt function from start (worker somewhere) to finish (algo somewhere), but I'm very unsure what that function really is and what it expects to happen.. e.g.
https://github.com/valhalla/valhalla/blob/efff65d35c68c9198d504e07087d3ae2b71e5e1f/src/thor/costmatrix.cc#L254-L257

not asking for enlightenment at this point, I'll dive a bit soon

**Comment by kevinkreiser** [comment]
that call will throw if the request should be aborted. then you'll see it's caught in the main worker function that started all the processing. it's nothing more complicated than that

**Comment by mkandulavm** [comment]
Hi
While I was on master, the code is not latest and is missing the interrupt block of code. I am sorry.
I am getting the latest and merging. Will revert on the interrupt shortly and confirm if it works and close the thread.

**Comment by mkandulavm** [comment]
Hi.

With the latest code , the interrupt is working. Thank you for the help and sorry for not checking thoroughly enough.
